python/cudf/cudf/core/dataframe.py

```diff
@@ -6145,6 +6145,45 @@ def __dataframe__(
             self, nan_as_null=nan_as_null, allow_copy=allow_copy
         )
 
+    def nunique(self, axis=0, dropna=True):
+        """
+        Count number of distinct elements in specified axis.
+        Return Series with number of distinct elements. Can ignore NaN values.
+        
+        Parameters
+        ----------
+        axis : {0 or 'index', 1 or 'columns'}, default 0
+            The axis to use. 0 or 'index' for row-wise, 1 or 'columns' for column-wise.
+        dropna : bool, default True
+            Don't include NaN in the counts.
+
+        Returns
+        -------
+        Series
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame({'A': [4, 5, 6], 'B': [4, 1, 1]})
+        >>> df.nunique()
+        A    3
+        B    2
+        dtype: int64
+        """
+        if dropna != True:
+            raise NotImplementedError(
+                "dropna parameter is not supported yet."
+            )
+
+        if axis != 0:
+            raise NotImplementedError(
+                "axis parameter is not supported yet."
+            )
+
+        return Series._from_data(
+            {None: [self[col].nunique() for col in self._data.names]},
+            as_index(self._data.names),
+        )
```

Comment:
I appreciate the use of `_from_data` here, but in this case I think the performance characteristics of this approach won't be all that different from calling the Series constructor directly with a dict since you're also calling `DataFrame.__getitem__`, constructing a Series there, and calling the Series's nunique method, not to mention building an index from host data. As a result, I think we may as well choose a less duplicative approach. Try implementing `nunique` in `Frame` using something like `return {name: col.distinct_count(method=method, dropna=dropna) for name, col in self._data.items()}`. That will return a dictionary. Then, in `DataFrame` you can do `return cudf.Series(super().nunique(method=method, dropna=dropna))`, and in `SingleColumnFrame` you can do `return sum(super().nunique(method=method, dropna=dropna).values()` (or any way of just getting the first value out since there's only one value in the dictionary for a single-columned frame). That will allow you to reuse the same implementation for Series and most Index types.

MultiIndex likely requires a different implementation. I'd guess that neither your current implementation nor the one that I've proposed will work in that case because we'll need a table-wide uniqueness check rather than a per-column measure like DataFrame uses. If there isn't an obvious way to implement that right now it's fine to skip it in this PR.

Reply:
Thank you very much for the information.